### ISSUE TITLE
Trino 356 : Deletion of partitions is not happening

### ISSUE THREAD

**Issue by vinay-kl** [trinodb/trino#8067]
>Overview
The partitions are not getting dropped when the delete query matches the partition referenced from a subquery.

> we are using `356` for the QA env where this issue is persistent, whereas on prod which is using `347` the issue is not.

```
trino> create table pso_test.par_and_buck_by1 (a varchar, b int, c varchar) WITH ( format = 'ORC', partitioned_by = ARRAY['c'], bucketed_by = ARRAY['a'], bucket_count = 10, transactional=true);
CREATE TABLE
```

```
trino> insert into pso_test.par_and_buck_by1 values ('a',1,'a'), ('b',2,'b');
INSERT: 2 rows
```

```
trino> select * from pso_test.par_and_buck_by1;
 a | b | c 
---+---+---
 b | 2 | b 
 a | 1 | a 
(2 rows)
```

```
trino> select * from pso_test."par_and_buck_by1$partitions";
 c 
---
 a 
 b 
(2 rows)
```

```
trino> delete from pso_test.par_and_buck_by1 where c in ('a','b');
DELETE: 2 rows

or 

trino> delete from pso_test.par_and_buck_by1 where c ='a' or c='b';
DELETE: 2 rows
```

```
trino> select * from pso_test."par_and_buck_by1$partitions";
 c 
---
 a 
 b 
(2 rows)
```

**Comment by vinay-kl** [comment]
also facing one more error case when tried to delete partitions from referencing the values selected from a different table in a subquery

```
trino> delete from pso_test.par_and_buck_by1 where c in (select distinct c from pso_test.par_and_buck_by_ext);
Query 20210525_115847_00051_3irbv failed: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
java.lang.IllegalArgumentException: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.findTableScanHandle(BeginTableWrite.java:259)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.findTableScanHandle(BeginTableWrite.java:251)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.findTableScanHandle(BeginTableWrite.java:248)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.findTableScanHandle(BeginTableWrite.java:245)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.findTableScanHandle(BeginTableWrite.java:248)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.getWriterTarget(BeginTableWrite.java:181)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.getWriterTarget(BeginTableWrite.java:196)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.getWriterTarget(BeginTableWrite.java:196)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:158)
	at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:78)
	at io.trino.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:84)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:73)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:74)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:38)
	at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:22)
	at io.trino.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
	at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:32)
	at io.trino.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:75)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:208)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:197)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:192)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:482)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:462)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:405)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_356____20210525_100723_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by hashhar** [comment]
@vinay-kl Instead of the last query `trino> select * from pso_test."par_and_buck_by1$partitions";` which lists partitions can you check if the data was deleted via `trino> select * from pso_test.par_and_buck_by1`?

**Comment by findepi** [comment]
> also facing one more error case when tried to delete partitions from referencing the values selected from a different table in a subquery
> 
> ```
> trino> delete from pso_test.par_and_buck_by1 where c in (select distinct c from pso_test.par_and_buck_by_ext);
> Query 20210525_115847_00051_3irbv failed: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
> java.lang.IllegalArgumentException: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
> ```

@vinay-kl thanks for reporting this! can you please open a separate issue for that bug?

**Comment by vinay-kl** [comment]
> @vinay-kl Instead of the last query `trino> select * from pso_test."par_and_buck_by1$partitions";` which lists partitions can you check if the data was deleted via `trino> select * from pso_test.par_and_buck_by1`?
> 
> cc: @djsstarburst

yes @hashhar indeed the data was deleted, checked the directory listing as well, delete delta is being created instead of drop call to HMS.

**Comment by vinay-kl** [comment]
> > also facing one more error case when tried to delete partitions from referencing the values selected from a different table in a subquery
> > ```
> > trino> delete from pso_test.par_and_buck_by1 where c in (select distinct c from pso_test.par_and_buck_by_ext);
> > Query 20210525_115847_00051_3irbv failed: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
> > java.lang.IllegalArgumentException: Invalid descendant for DeleteNode or UpdateNode: io.trino.sql.planner.plan.ExchangeNode
> > ```
> 
> @vinay-kl thanks for reporting this! can you please open a separate issue for that bug?

sure @findepi.

**Comment by electrum** [comment]
If you run the same `DELETE` command in Hive, does it also remove the partitions?

We explicitly changed the behavior of `DELETE` for transactional tables in #7621 to always do a row-by-row delete, rather than dropping the partitions. This is because dropping a partition is not transactional and would break ACID semantics for concurrent readers.

This leaves open the question of how to drop partitions for a transactional table. We could add an explicit `system.drop_partition` procedure, or add a session property to restore the old behavior of having `DELETE` drop partitions rather than row-by-row delete.

**Comment by electrum** [comment]
I realized we missed this in the release notes: https://github.com/trinodb/trino/pull/8072

**Comment by vinay-kl** [comment]
> If you run the same `DELETE` command in Hive, does it also remove the partitions?
> 
> We explicitly changed the behavior of `DELETE` for transactional tables in #7621 to always do a row-by-row delete, rather than dropping the partitions. This is because dropping a partition is not transactional and would break ACID semantics for concurrent readers.
> 
> This leaves open the question of how to drop partitions for a transactional table. We could add an explicit `system.drop_partition` procedure, or add a session property to restore the old behavior of having `DELETE` drop partitions rather than row-by-row delete.

@electrum `hive command does not drop partitions even when it matches the whole`. 

we wanted to overwrite the existing partitions but as trino errors out with `DIRECT_TO_TARGET_EXISTING_DIRECTORY` when used on non-HDFS and we are using Azure Gen2.

so we thought we'd remove partitions before we can re-insert by setting `default insert as overwrite` and there'd be no need to perform compaction laterwards.

also, another question on the similar lines, why not create a `new base dir with latest write_id when overwrite is selected`?

**Comment by electrum** [comment]
> why not create a new base dir with latest write_id when it is overwrite, to match with the hive behavior?

Do you mean when you run `INSERT OVERWRITE` in Hive against a transactional table? If so, then it sounds like we could do that as well when using `insert_existing_partitions_behavior=OVERWRITE` in Trino.

Could you file a new issue for this, with an example of what you run in Hive and what the directory looks like before/after the overwrite? Hive transactional behavior is not documented, so we have to reverse engineer it, and having these examples is a good place to start, and will help when writing test cases.

**Comment by electrum** [comment]
For the issue of restoring the old "drop partition" behavior when running `DELETE` on transactional tables, we could add a config+session property to the Hive connector. It would go in `applyDelete()` to return `Optional.of(handle)` when set: https://github.com/trinodb/trino/commit/a9478d091d4a6519e3befddd86fcc1022c691a03

**Comment by electrum** [comment]
@vinay-kl Regarding your use case, do you actually need transactional tables? Do you ever do row level deletes or take advantage of the ACID semantics? If not, then making the tables non-transactional should solve your problem.

**Comment by vinay-kl** [comment]
> @vinay-kl Regarding your use case, do you actually need transactional tables? Do you ever do row level deletes or take advantage of the ACID semantics? If not, then making the tables non-transactional should solve your problem.

thanks, @electrum and yes we'd be going ahead with an external table approach instead for daily overwrite workloads.

**Comment by vinay-kl** [comment]
> > why not create a new base dir with latest write_id when it is overwrite, to match with the hive behavior?
> 
> Do you mean when you run `INSERT OVERWRITE` in Hive against a transactional table? If so, then it sounds like we could do that as well when using `insert_existing_partitions_behavior=OVERWRITE` in Trino.
> 
> Could you file a new issue for this, with an example of what you run in Hive and what the directory looks like before/after the overwrite? Hive transactional behavior is not documented, so we have to reverse engineer it, and having these examples is a good place to start, and will help when writing test cases.

sure @electrum will create another case with detailed info, thanks..!

**Comment by longtran84** [comment]
I got error as :

trino> insert into minio.kai_data_management.par_and_buck_by1 values ('a',1,'a'), ('b',2,'b');
Query 20210822_110300_00634_t37mp failed: Internal error
java.lang.UnsupportedOperationException
        at io.trino.plugin.hive.metastore.HiveMetastore.openTransaction(HiveMetastore.java:142)
        at io.trino.plugin.hive.metastore.cache.CachingHiveMetastore.openTransaction(CachingHiveMetastore.java:925)
        at io.trino.plugin.hive.HiveMetastoreClosure.openTransaction(HiveMetastoreClosure.java:300)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.makeHiveTransaction(SemiTransactionalHiveMetastore.java:1264)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.beginOperation(SemiTransactionalHiveMetastore.java:1238)
        at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.beginInsert(SemiTransactionalHiveMetastore.java:1216)
        at io.trino.plugin.hive.HiveMetadata.beginInsert(HiveMetadata.java:1689)
        at io.trino.plugin.hive.HiveMetadata.beginInsert(HiveMetadata.java:286)
        at io.trino.spi.connector.ConnectorMetadata.beginInsert(ConnectorMetadata.java:473)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.beginInsert(ClassLoaderSafeConnectorMetadata.java:454)
        at io.trino.metadata.MetadataManager.beginInsert(MetadataManager.java:906)
        at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.createWriterTarget(BeginTableWrite.java:227)
        at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:174)
        at io.trino.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:93)
        at io.trino.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
        at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:84)
        at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:73)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
        at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:74)
        at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:38)
        at io.trino.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:22)
        at io.trino.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
        at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:83)
        at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:32)
        at io.trino.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:78)
        at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:216)
        at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:201)
        at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:196)
        at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:488)
        at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:468)
        at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:411)
        at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
        at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
        at io.trino.$gen.Trino_358____20210813_023135_2.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
        Suppressed: java.lang.Exception: Current plan:
                Output[rows]
                │   Layout: [rows:bigint]
                └─ TableCommit[minio:kai_data_management:par_and_buck_by1 buckets=10]
                   │   Layout: [rows:bigint]
                   └─ LocalExchange[SINGLE] ()
                      │   Layout: [partialrows:bigint, fragment:varbinary]
                      └─ RemoteExchange[GATHER]
                         │   Layout: [partialrows:bigint, fragment:varbinary]
                         └─ TableWriter
                            │   Layout: [partialrows:bigint, fragment:varbinary]
                            │   a := field
                            │   b := field_0
                            │   c := field_1
                            └─ LocalExchange[SINGLE] ()
                               │   Layout: [field:varchar(1), field_0:integer, field_1:varchar(1)]
                               └─ RemoteExchange[REPARTITION]
                                  │   Layout: [field:varchar(1), field_0:integer, field_1:varchar(1)]
                                  └─ Values
                                         Layout: [field:varchar(1), field_0:integer, field_1:varchar(1)]
                                         ('a', 1, 'a')
                                         ('b', 2, 'b')

                at io.trino.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:84)
                ... 12 more


please help me , thanks

**Comment by jerryleooo** [comment]
Just tested that this issue still persists on version 389

**Comment by findepi** [comment]
Should have been fixed by @djsstarburst 's https://github.com/trinodb/trino/pull/13926.

Please file a new issue if it is not the case.
